### DISCUSSION TITLE
How to Push and Install a local APK

### DISCUSSION THREAD

**Discussion by JinalShah2002** [yume-chan/ya-webadb#575] [Q&A]
Hi! 
I am using this library for a personal project and wanted to know how I can push and install an APK to a connected Android device. For context, I have an APK file that is stored in my local file system. I would like to push and install this APK to my Android device using the library, Javascript, and React. Any insight would be appreciated. 

@yume-chan

**Comment by yume-chan** [comment]
See https://github.com/yume-chan/ya-webadb/blob/b066ca01e7c6d17e02c969db94c458f9cd5d0463/apps/demo/src/pages/install.tsx

**Reply by JinalShah2002** [reply]
Thank you! Do you know how I can fetch the APK locally in code? I want it so that from a click of a button the APK automatically downloads on the connected Android device.

**Reply by JinalShah2002** [reply]
I would also like to mention that I want the application to be able to push and install the APK on the device. The user should click a button and then the application will automatically push and install the APK on the connected device. For this, I need to somehow fetch the APK locally and then push and install it on the device. Any insight on this is also appreciated.

**Reply by yume-chan** [reply]
You can't read local file system from Web. You can let the user pick a file (your APK in this case) after clicking a button. That's exactly what my above link demonstrates. 

You can also download an APK file from your server and install it onto the connected device, just replace the `pickFile` call with a `fetch`.

**Reply by JinalShah2002** [reply]
Ok, so I can't hard code the file path and have the application pull it? For example, if my APK was located in the apks_folder within my project directory, I can't have JS pull the APK?

If possible, could you provide pseudocode or an example of how you would use `fetch` as a replacement for `pickFile`? Would like to see how the APK push and install process would happen using `fetch` if you don't mind.

**Reply by yume-chan** [reply]
You can include the APK into your project. As a Web project, all files are hosted on your server, so `fetch` is used to fetch that.

```diff
     install = async () => {
-        const file = await pickFile({ accept: ".apk" });
-        if (!file) {
+        const response = await fetch("https://www.example.com/my.apk");
+        if (!response.ok) {
             return;
````

```diff
         const pm = new PackageManager(GLOBAL_STATE.adb!);
-            file.size,
-            createFileStream(file)
+            +response.headers.get("Content-Length"),
+            response.body
                 .pipeThrough(new WrapConsumableStream())
````
